File: R/prior_strength.R
# Experimental helpers for prior strength calculations.

PRIOR_STRENGTH_SPECS <- list(
  normal_mean = list(
    label = "Normal mean (known sigma)",
    prior_params = c("mu0", "sigma0"),
    likelihood_params = c("sigma"),
    positive_params = c("sigma0", "sigma"),
    notes = "Posterior mean uses precision weights; assumes known observation SD.",
    weight_fn = function(prior, likelihood, n) {
      list(
        prior = 1 / (prior$sigma0^2),
        likelihood = n / (likelihood$sigma^2)
      )
    },
    n_equal_fn = function(prior, likelihood) {
      (likelihood$sigma^2) / (prior$sigma0^2)
    }
  ),
  bernoulli = list(
    label = "Bernoulli/Binomial (Beta prior)",
    prior_params = c("alpha", "beta"),
    likelihood_params = character(0),
    positive_params = c("alpha", "beta"),
    notes = "Posterior mean is weighted by prior pseudo-counts (alpha+beta) vs. n.",
    weight_fn = function(prior, likelihood, n) {
      list(
        prior = prior$alpha + prior$beta,
        likelihood = n
      )
    },
    n_equal_fn = function(prior, likelihood) {
      prior$alpha + prior$beta
    }
  ),
  poisson = list(
    label = "Poisson (Gamma prior on rate)",
    prior_params = c("alpha", "beta"),
    likelihood_params = character(0),
    positive_params = c("alpha", "beta"),
    notes = "Posterior mean uses weights beta (prior) vs. n (likelihood).",
    weight_fn = function(prior, likelihood, n) {
      list(
        prior = prior$beta,
        likelihood = n
      )
    },
    n_equal_fn = function(prior, likelihood) {
      prior$beta
    }
  )
)

prior_strength_choices <- function() {
  labels <- vapply(names(PRIOR_STRENGTH_SPECS), function(x) PRIOR_STRENGTH_SPECS[[x]]$label, character(1))
  setNames(names(PRIOR_STRENGTH_SPECS), labels)
}

prior_strength_spec <- function(likelihood) {
  spec <- PRIOR_STRENGTH_SPECS[[likelihood]]
  if (is.null(spec)) {
    stop("Unknown likelihood: ", likelihood)
  }
  spec
}

ensure_named_params <- function(params, required, label) {
  if (is.null(params)) {
    params <- list()
  }
  if (!is.list(params)) {
    stop(label, " parameters must be a list.")
  }
  missing <- setdiff(required, names(params))
  if (length(missing) > 0) {
    stop(label, " parameters missing: ", paste(missing, collapse = ", "))
  }
  params
}

ensure_positive_params <- function(params, names, label) {
  for (nm in names) {
    val <- params[[nm]]
    if (!is.numeric(val) || length(val) != 1 || is.na(val)) {
      stop(label, " parameter '", nm, "' must be a single numeric value.")
    }
    if (val <= 0) {
      stop(label, " parameter '", nm, "' must be > 0.")
    }
  }
}

ensure_positive_n <- function(n) {
  if (!is.numeric(n) || length(n) != 1 || is.na(n)) {
    stop("n must be a single numeric value.")
  }
  if (n <= 0) {
    stop("n must be > 0.")
  }
}

prior_strength_weights <- function(likelihood, prior_params, likelihood_params, n) {
  spec <- prior_strength_spec(likelihood)
  prior_params <- ensure_named_params(prior_params, spec$prior_params, "Prior")
  likelihood_params <- ensure_named_params(likelihood_params, spec$likelihood_params, "Likelihood")
  ensure_positive_params(prior_params, spec$positive_params, "Prior")
  ensure_positive_params(likelihood_params, spec$positive_params, "Likelihood")
  ensure_positive_n(n)

  weights <- spec$weight_fn(prior_params, likelihood_params, n)
  total <- weights$prior + weights$likelihood
  list(
    prior_weight = weights$prior,
    likelihood_weight = weights$likelihood,
    prior_share = weights$prior / total,
    likelihood_share = weights$likelihood / total
  )
}

prior_strength_thresholds <- function(likelihood, prior_params, likelihood_params, n_is_integer = TRUE) {
  spec <- prior_strength_spec(likelihood)
  prior_params <- ensure_named_params(prior_params, spec$prior_params, "Prior")
  likelihood_params <- ensure_named_params(likelihood_params, spec$likelihood_params, "Likelihood")
  ensure_positive_params(prior_params, spec$positive_params, "Prior")
  ensure_positive_params(likelihood_params, spec$positive_params, "Likelihood")

  n_equal <- spec$n_equal_fn(prior_params, likelihood_params)
  if (n_is_integer) {
    list(
      n_equal = n_equal,
      n_min_likelihood_dominates = floor(n_equal) + 1,
      n_max_prior_dominates = ceiling(n_equal) - 1
    )
  } else {
    list(
      n_equal = n_equal,
      n_min_likelihood_dominates = n_equal,
      n_max_prior_dominates = n_equal
    )
  }
}

prior_strength_summary <- function(likelihood, prior_params, likelihood_params, n, n_is_integer = TRUE) {
  spec <- prior_strength_spec(likelihood)
  weights <- prior_strength_weights(likelihood, prior_params, likelihood_params, n)
  thresholds <- prior_strength_thresholds(likelihood, prior_params, likelihood_params, n_is_integer = n_is_integer)

  dominance <- if (weights$prior_weight > weights$likelihood_weight) {
    "prior"
  } else if (weights$prior_weight < weights$likelihood_weight) {
    "likelihood"
  } else {
    "tie"
  }

  data.frame(
    likelihood = likelihood,
    label = spec$label,
    n = n,
    prior_weight = weights$prior_weight,
    likelihood_weight = weights$likelihood_weight,
    prior_share = weights$prior_share,
    likelihood_share = weights$likelihood_share,
    dominance = dominance,
    n_equal = thresholds$n_equal,
    n_min_likelihood_dominates = thresholds$n_min_likelihood_dominates,
    n_max_prior_dominates = thresholds$n_max_prior_dominates,
    notes = spec$notes,
    stringsAsFactors = FALSE
  )
}

File: web/js/priorStrength.js
// Conjugate prior-vs-likelihood weight math. Mirrors R/prior_strength.R,
// including the fix where positive-parameter validation checks each parameter
// set only against the names it actually owns.

export const PRIOR_STRENGTH_SPECS = {
  normal_mean: {
    label: "Normal mean (known sigma)",
    priorParams: ["mu0", "sigma0"],
    likelihoodParams: ["sigma"],
    positiveParams: ["sigma0", "sigma"],
    notes: "Posterior mean uses precision weights; assumes known observation SD.",
    weightFn: (prior, likelihood, n) => ({
      prior: 1 / (prior.sigma0 * prior.sigma0),
      likelihood: n / (likelihood.sigma * likelihood.sigma),
    }),
    nEqualFn: (prior, likelihood) =>
      (likelihood.sigma * likelihood.sigma) / (prior.sigma0 * prior.sigma0),
  },
  bernoulli: {
    label: "Bernoulli/Binomial (Beta prior)",
    priorParams: ["alpha", "beta"],
    likelihoodParams: [],
    positiveParams: ["alpha", "beta"],
    notes: "Posterior mean is weighted by prior pseudo-counts (alpha+beta) vs. n.",
    weightFn: (prior, likelihood, n) => ({
      prior: prior.alpha + prior.beta,
      likelihood: n,
    }),
    nEqualFn: (prior) => prior.alpha + prior.beta,
  },
  poisson: {
    label: "Poisson (Gamma prior on rate)",
    priorParams: ["alpha", "beta"],
    likelihoodParams: [],
    positiveParams: ["alpha", "beta"],
    notes: "Posterior mean uses weights beta (prior) vs. n (likelihood).",
    weightFn: (prior, likelihood, n) => ({
      prior: prior.beta,
      likelihood: n,
    }),
    nEqualFn: (prior) => prior.beta,
  },
};

export function priorStrengthSpec(likelihood) {
  const spec = PRIOR_STRENGTH_SPECS[likelihood];
  if (!spec) {
    throw new Error(`Unknown likelihood: ${likelihood}`);
  }
  return spec;
}

function ensureNamedParams(params, required, label) {
  const p = params || {};
  const missing = required.filter((name) => p[name] === undefined);
  if (missing.length > 0) {
    throw new Error(`${label} parameters missing: ${missing.join(", ")}`);
  }
  return p;
}

function ensurePositiveParams(params, names, label) {
  for (const name of names) {
    const value = params[name];
    if (typeof value !== "number" || Number.isNaN(value)) {
      throw new Error(`${label} parameter '${name}' must be a single numeric value.`);
    }
    if (value <= 0) {
      throw new Error(`${label} parameter '${name}' must be > 0.`);
    }
  }
}

function ensurePositiveN(n) {
  if (typeof n !== "number" || Number.isNaN(n)) {
    throw new Error("n must be a single numeric value.");
  }
  if (n <= 0) {
    throw new Error("n must be > 0.");
  }
}

function validate(spec, priorParams, likelihoodParams) {
  const prior = ensureNamedParams(priorParams, spec.priorParams, "Prior");
  const likelihood = ensureNamedParams(likelihoodParams, spec.likelihoodParams, "Likelihood");
  ensurePositiveParams(
    prior,
    spec.positiveParams.filter((name) => spec.priorParams.includes(name)),
    "Prior"
  );
  ensurePositiveParams(
    likelihood,
    spec.positiveParams.filter((name) => spec.likelihoodParams.includes(name)),
    "Likelihood"
  );
  return { prior, likelihood };
}

export function priorStrengthWeights(likelihood, priorParams, likelihoodParams, n) {
  const spec = priorStrengthSpec(likelihood);
  const params = validate(spec, priorParams, likelihoodParams);
  ensurePositiveN(n);

  const weights = spec.weightFn(params.prior, params.likelihood, n);
  const total = weights.prior + weights.likelihood;
  return {
    priorWeight: weights.prior,
    likelihoodWeight: weights.likelihood,
    priorShare: weights.prior / total,
    likelihoodShare: weights.likelihood / total,
  };
}

export function priorStrengthThresholds(likelihood, priorParams, likelihoodParams, nIsInteger = true) {
  const spec = priorStrengthSpec(likelihood);
  const params = validate(spec, priorParams, likelihoodParams);

  const nEqual = spec.nEqualFn(params.prior, params.likelihood);
  if (nIsInteger) {
    return {
      nEqual,
      nMinLikelihoodDominates: Math.floor(nEqual) + 1,
      nMaxPriorDominates: Math.ceil(nEqual) - 1,
    };
  }
  return {
    nEqual,
    nMinLikelihoodDominates: nEqual,
    nMaxPriorDominates: nEqual,
  };
}

// Closed-form posterior metrics for the Normal-mean case, shared by the grid
// (plots) and the summary table so the two can never drift apart.
export function normalMeanMetrics(mu0, tau0, betaHat, sigma, n) {
  const wPrior = 1 / (tau0 * tau0);
  const wLike = n / (sigma * sigma);
  const s2 = (sigma * sigma) / n;
  return {
    priorShare: wPrior / (wPrior + wLike),
    posteriorMean: (wPrior * mu0 + wLike * betaHat) / (wPrior + wLike),
    kl: 0.5 * (Math.log(s2 / (tau0 * tau0)) + (tau0 * tau0 + (mu0 - betaHat) ** 2) / s2 - 1),
  };
}
